### ISSUE TITLE
After trained how to use the features?

### ISSUE THREAD

**Issue by ConleyKong** [Chosko/img-face-detection#1]
Hi ,I have run your code on my matlab, it works very good ,I appreciate your masterpiece very well ,but I meet a problem , after trained my features how to use them into my face detection process? By the way,  I'm a beginner of machine learning ,sorry to bother you……

**Comment by Chosko** [comment]
Hi @ConleyKong. To see how to use the features, take a look at the file detect.m. It's a function that detects the faces in an image adding rectangles around them.

```
    function Y = detect(windowsize, X, detector, progress, scale, factor)
```
- windowsize is the size of the window used to train the classifiers
- X is the input image
- detector is the trained cascade of classifiers. You can find it in the variable "cascade" when you load TRAINING_OUT_FILE (see config_training.m) 
- progress should tell the function to print a progress bar
- scale (optional), if present, forces the detection at a fixed scale of the windowsize
- factor (optional) may be left to 1, it's the factor used in classifyimg.m

However, I must warn you about the accuracy of this project. It works, but the training set present in this repository is absolutely inadequate to provide a good detection.

I just added "Relazione.pdf" which is the document I presented to the professors of my University. Although it's in Italian, I think you should be able to understand the charts and the results at the end of the document. Basically the Algorithm is OK, but the training set is very poor (especially in the number of negatives) and the size of the pictures is too small to train a good cascade of selectors. In the end, if you try to catch a face in a real image using the classifiers trained with this set, the result will be a lot of false positives.

I write a short summary in English for you, to explain better the situation:

The main differences between my project (MP) and the original project of Viola and Jones (VJ) are:
- Size of the images in the training set: 19x19 pixel (MP) vs 24x24 pixel (VJ)
- Number of images with faces in training set (positives): 472 (MP) vs 4916 (VJ)
- Number of images/subwindows without faces in training set (negatives): 23573 (MP) vs 350mln (VJ)
- Number of weak classifiers in the final detector: 160 (MP) vs 6060 (VJ)
- False positive rate on a test set: 0.0446 (MP) vs 10e-6 (suggested by VJ in their paper)

The last values are interesting. Let's say you want to detect a face in an image of 512x512 px. The total number of subwindows is N=20573.  With a false positive rate of F1=0.0446 (MP), you will find about NxF1=917 false detections, while the real face is just one. With a  false positive rate of F2=10e-6 (VJ) the expect number of false detections is NxF2=0,020573, which is < 1, so you expect not to have false positives but only the real face detected. You can see this graphically at the end of "Relazione.pdf": the image is the one full of white rectangles.

All these values depends only from the training set. My training set is poor because I made this project for the University and I didn't have much time to spend in choosing the images.

If you want to use this project in production, I strongly suggest you to take the time to create a custom training set with at least the number of images Viola and Jones used in their original paper.
And after that, consider that the time needed to train the classifier will raise a lot.

Cheers!

**Comment by ConleyKong** [comment]
Hi @Chosko Chosko, very appreciate that you can reply my question, according to your description I have got a rather clear understanding about the program ,no words can describe my gratitude in my heart, hope next time I can learn more from you!         Thank you !
